### PR TITLE
embed: support registering client handlers

### PR DESCRIPTION
**Pull request by siddontang**
@xiang90 

This is for https://github.com/coreos/etcd/issues/6014

### CODE REVIEW

File: embed/config.go [hunk 1]
```diff
@@ -98,6 +99,10 @@ type Config struct {
 
 	// ForceNewCluster starts a new cluster even if previously started; unsafe.
 	ForceNewCluster bool `json:"force-new-cluster"`
+
+	// ClientHandlers is for registering own handlers.
+	// You must ensure the route paths can't be conflicted with etcd's.
+	ClientHandlers map[string]http.Handler `json:"-"`
```
Comment (xiang90) [inline embed/config.go]: AdditionalHandlers map[string]http.Handler?

We should mention that the key in the map is the path for the handler?

File: embed/serve.go [hunk 2]
```diff
@@ -42,6 +42,8 @@ type serveCtx struct {
 
 	ctx    context.Context
 	cancel context.CancelFunc
+
+	handlers map[string]http.Handler
```
Comment (xiang90) [inline embed/serve.go]: additionalHandlers?

File: embed/serve.go [hunk 3]
```diff
@@ -170,3 +170,14 @@ func (sctx *serveCtx) registerGateway(opts []grpc.DialOption) (*gw.ServeMux, err
 	}
 	return gwmux, nil
 }
+
+func (sctx *serveCtx) createMux(gwmux *gw.ServeMux, handler http.Handler) *http.ServeMux {
+	httpmux := http.NewServeMux()
+	for name, h := range sctx.handlers {
```
Comment (xiang90) [inline embed/serve.go]: path, h

**Comment by xiang90** [comment]
defer to @heyitsanthony

File: embed/config.go (hunk 1 above)
Comment (xiang90) [inline embed/config.go]: Also mention that this is only designed for embedding etcd into other applications? So we ignore its json representation?

File: embed/serve.go (hunk 2 above)
Comment (heyitsanthony) [inline embed/serve.go]: `userHandlers`?

File: embed/config.go (hunk 1 above)
Comment (heyitsanthony) [inline embed/config.go]: `UserHandlers`?

**Comment by heyitsanthony** [comment]
lgtm aside from the `s/name/path` bit @xiang90 suggested. I'm not too picky about the naming; `ClientHandlers`/`clientHandlers` is probably fine since it fits the client/peer listener naming scheme.

**Comment by xiang90** [comment]
@heyitsanthony UserHandlers seem to be better. Client is confusing when you think about etcd clients.

**Comment by heyitsanthony** [comment]
@xiang90 ok, no strong opinion

**Comment by siddontang** [comment]
PTAL

File: embed/serve.go (hunk 2 above)
Comment (xiang90) [inline embed/serve.go]: update this? @siddontang

**Comment by xiang90** [comment]
LGTM after addressing https://github.com/coreos/etcd/pull/6063#discussion_r72562720. Also please squash your commits into one.

**Comment by xiang90** [comment]
lgtm. please squash commits.

**Pull request merged**